internal/logstreamer/broker.go

```diff
@@ -0,0 +1,66 @@
+package logstreamer
+
+type BrokerSub struct {
+	load chan []byte
+}
+
+// Broker is message broker consuming messages from one
+// source and delivering them to multiple subscribers
+type Broker struct {
+	listener chan []byte
+	sub      chan *BrokerSub
+	unsub    chan *BrokerSub
+	quit     chan struct{}
+}
+
+func NewBroker() *Broker {
+	return &Broker{
+		listener: make(chan []byte, 1),
+		sub:      make(chan *BrokerSub, 1),
+		unsub:    make(chan *BrokerSub, 1),
+		quit:     make(chan struct{}),
+	}
+}
+
+func (b *Broker) Subscribe() *BrokerSub {
+	s := &BrokerSub{
+		load: make(chan []byte),
+	}
+	b.sub <- s
+	return s
+}
+
+func (b *Broker) Unsubscribe(s *BrokerSub) {
+	b.unsub <- s
+}
+
+func (b *Broker) Send(msg []byte) {
+	b.listener <- msg
+}
+
+func (b *Broker) Stop() {
+	close(b.quit)
+}
+
+func (b *Broker) Broadcast() {
+	subs := map[*BrokerSub]struct{}{}
+	for {
+		select {
+		case msg := <-b.listener:
+			// Resending the published message to all the subs
+			for sub := range subs {
+				sub.load <- msg
+			}
+		case sub := <-b.sub:
+			subs[sub] = struct{}{}
+		case sub := <-b.unsub:
+			close(sub.load)
+			delete(subs, sub)
```

Comment:
It seemed like a nice idea not to introduce an additional locking mechanism and handle everything with this kind of "out of the box" channels sync, but you're right, it's better to manage `subs` separately in order to prevent the possible edge cases